File: 166 Fraction to Recurring Decimal/166 Fraction to Recurring Decimal.py
"""
Given two integers representing the numerator and denominator of a fraction, return the fraction in string format.
If the fractional part is repeating, enclose the repeating part in parentheses.
If multiple answers are possible, return any of them.
It is guaranteed that the length of the answer string is less than 104 for all the given inputs.

Example 1:

Input: numerator = 1, denominator = 2
Output: "0.5"

Example 2:

Input: numerator = 2, denominator = 1
Output: "2"

Example 3:

Input: numerator = 2, denominator = 3
Output: "0.(6)"

Example 4:

Input: numerator = 4, denominator = 333
Output: "0.(012)"

Example 5:

Input: numerator = 1, denominator = 5
Output: "0.2"
"""
class Solution:
    def fractionToDecimal(self, numerator: int, denominator: int) -> str:
        if numerator == 0:
            return '0'
        if numerator < 0 and denominator < 0:
            return self.fractionToDecimal(abs(numerator), abs(denominator))
        elif numerator < 0 or denominator < 0:
            return '-' + self.fractionToDecimal(abs(numerator), abs(denominator))
        div, mod = numerator // denominator, numerator % denominator
        curr = str(div)
        if mod == 0:
            return curr
        subZero = '.'
        mapping = {}
        while True:
            mod *= 10
            temp = mod
            div, mod = mod // denominator, mod % denominator
            if mod == 0:
                subZero += str(div)
                return curr + subZero
            if temp in mapping:
                repeatIndex = mapping[temp]
                return curr + subZero[:repeatIndex] + '(' + subZero[repeatIndex:] + ')'
            else:
                mapping[temp] = len(subZero)
                subZero += str(div)

File: 166 Fraction to Recurring Decimal/166 Fraction to Recurring Decimal.go
import "fmt"

func fractionToDecimal(numerator int, denominator int) string {

    if numerator == 0 {
        return "0"
    }
    if numerator == denominator {
        return "1"
    }

    sign := 1

    if numerator < 0 {
        numerator *= -1
        sign *= -1
    }

    if denominator < 0 {
        denominator *= -1
        sign *= -1
    }

    s := ""
    if sign < 0 {
        s += "-"
    }

    a := numerator/denominator
    s += fmt.Sprintf("%d", a)

    rem := numerator % denominator
    if rem > 0 {
        s += "."
    }

    index := len(s)
    ok := false
    m := make(map[int]int)

    for rem != 0 {

        index, ok = m[rem]
        if !ok {
            m[rem] = len(s)
        } else {
            s = s[0:index] + "(" + s[index:] + ")"
            break
        }

        rem *= 10

        a = rem/denominator
        rem = rem % denominator
        s += fmt.Sprintf("%d", a)
    }

    return s
}
